pkg/controller/usersignup/usersignup_controller_test.go

```diff
@@ -813,6 +813,141 @@ func TestUserSignupWithInvalidNameNotOK(t *testing.T) {
+		},
+		Spec: v1alpha1.UserSignupSpec{
+			Username:    "foo#bar@redhat.com",
+			Approved:    false,
+			Deactivated: true,
+		},
+	}
+
+	r, req, _ := prepareReconcile(t, userSignup.Name, userSignup, configMap(config.UserApprovalPolicyAutomatic), basicNSTemplateTier)
+
+	_, err := r.Reconcile(req)
+	require.NoError(t, err)
+
+	key := types.NamespacedName{
+		Namespace: operatorNamespace,
+		Name:      userSignup.Name,
+	}
+	instance := &v1alpha1.UserSignup{}
+	err = r.client.Get(context.TODO(), key, instance)
+	require.NoError(t, err)
+
+	test.AssertConditionsMatch(t, instance.Status.Conditions,
+		v1alpha1.Condition{
+			Type:   v1alpha1.UserSignupComplete,
+			Status: v1.ConditionTrue,
+			Reason: "Deactivated",
+		},
+	)
+}
+
+func TestUserSignupDeactivatedAfterMURCreated(t *testing.T) {
+	userSignup := &v1alpha1.UserSignup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      uuid.NewV4().String(),
+			Namespace: operatorNamespace,
+			UID:       types.UID(uuid.NewV4().String()),
+		},
+		Spec: v1alpha1.UserSignupSpec{
+			Username: "john.doe@redhat.com",
+		},
+		Status: v1alpha1.UserSignupStatus{
+			Conditions: []v1alpha1.Condition{
+				{
+					Type:   v1alpha1.UserSignupComplete,
+					Status: v1.ConditionTrue,
+				},
+				{
+					Type:   v1alpha1.UserSignupApproved,
+					Status: v1.ConditionTrue,
+					Reason: "ApprovedAutomatically",
+				},
+			},
+			CompliantUsername: "john-doe-at-redhat-com",
+		},
+	}
+
+	mur := &v1alpha1.MasterUserRecord{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "john-doe-at-redhat-com",
+			Namespace: operatorNamespace,
+			UID:       types.UID(uuid.NewV4().String()),
+			Labels:    map[string]string{toolchainv1alpha1.MasterUserRecordUserIDLabelKey: userSignup.Name},
+		},
+	}
+
+	r, req, _ := prepareReconcile(t, userSignup.Name, userSignup, mur, configMap(config.UserApprovalPolicyAutomatic), basicNSTemplateTier)
+
+	// Lookup the UserSignup
+	key := types.NamespacedName{
+		Namespace: operatorNamespace,
+		Name:      userSignup.Name,
+	}
+	err := r.client.Get(context.TODO(), key, userSignup)
+	require.NoError(t, err)
+
+	// Now mark the UserSignup as Deactivated
+	userSignup.Spec.Deactivated = true
+	err = r.client.Update(context.TODO(), userSignup)
+	require.NoError(t, err)
+
+	// Reconcile
+	_, err = r.Reconcile(req)
+	require.NoError(t, err)
+
+	err = r.client.Get(context.TODO(), key, userSignup)
+	require.NoError(t, err)
+
+	// Confirm the status is now set to Deactivating
+	test.AssertConditionsMatch(t, userSignup.Status.Conditions,
+		v1alpha1.Condition{
+			Type:   v1alpha1.UserSignupApproved,
+			Status: v1.ConditionTrue,
+			Reason: "ApprovedAutomatically",
+		},
+		v1alpha1.Condition{
+			Type:   v1alpha1.UserSignupComplete,
+			Status: v1.ConditionFalse,
+			Reason: "Deactivating",
+		})
+
+	murs := &v1alpha1.MasterUserRecordList{}
+
+	// The MUR should have now been deleted
+	err = r.client.List(context.TODO(), murs)
+	require.NoError(t, err)
+	require.Len(t, murs.Items, 0)
```

Comment:
could you please cut the test here and verify the next reconcile in a new test? It would be much better if we had one unit test for one operation - it's much easier to read as well as debug in case of failures

Reply:
you should also verify a case when:
`UserSignup.Reason = Deactivating` but the MUR still exists - I guess that in that case, the client will return an error for another attempt of the MUR deletion - you should verify that use-case too.

Reply:
I'd prefer not to cut the test at this point.  Moving the remainder of the lines to a different test will still require all of the setup that comes before.  In fact, everything before the line `	userSignup.Spec.Deactivated = true` is essentially setup, and the remainder of the test is simply simulating the controller running in a live environment and validating that the resource states are correct.

Reply:
let me try to clarify it here - in your test you verify two things:
* for the first reconcile, that the MUR is deleted and the UserSignup status condition is set to `Deactivating`
* for the second reconcile, you verify that when there is no MUR present then the  UserSignup status condition is set to `Deactivated`

for me, these are two cases that should be tested separately and not together. In addition, setting `userSignup.Spec.Deactivated = true` doesn't have to be done via client call, but during the UserSignup initialization - there is no difference and it will make the test shorter and easier to read.

If you are afraid of doing the test setup again, you can use subtests and reuse the test setup.

Reply:
and please, don't forget to verify also the cases I mentioned in https://github.com/codeready-toolchain/host-operator/pull/126#discussion_r366989267

Reply:
@MatousJobanek I've made your suggested changes, thanks.  It seems a bit redundant to me though to use subtests for this.. the second subtest cannot exist in isolation without the previous subtest executing first.  The value of having the subtests at all then seems to be merely as a syntactic visual delineation between the sections, or am I missing something here?

Reply:
no, they are not dependent - both subtests can be executed either on their own or both of them together. There is nothing in the second subtest that would require anything from the first one. 
Try it out ;-)

Reply:
I've also added a test for when the condition status is set to `Deactivating` but the MUR still exists, in this case the controller simply deletes the MUR, as the workflow isn't based on the status value.

Reply:
Ah you're right, my apologies... I didn't take into account that `prepareReconcile` sets up the complete state of the resources for each test.

Reply:
> I've also added a test for when the condition status is set to Deactivating but the MUR still exists

thanks for adding this, but I was talking about covering the case when the `client.Delete` call returns an error.

Reply:
> > I've also added a test for when the condition status is set to Deactivating but the MUR still exists
> 
> thanks for adding this, but I was talking about covering the case when the `client.Delete` call returns an error.

Ah, that wasn't clear but no problem, I will add a test to validate the behaviour when the MUR delete fails.  In this case the reason won't be set to `Deactivating` though, instead we should expect a reason value of `UnableToDeleteMUR`.

Reply:
I've added this test now.

Reply:
Thanks a lot and sorry for not being clear :-)